warehouse-management-service/pkg/database/postgres/postgres.go

```diff
@@ -0,0 +1,36 @@
+package postgres
+
+import (
+	"database/sql"
+	"fmt"
+	"warehouse-management-service/pkg/database"
+
+	// Loads postgres drivers
+	_ "github.com/lib/pq"
+)
+
+type postgres struct {
+	db *sql.DB
+}
+
+func New(host, port, user, password, dbName string, sslMode bool) (database.Service, error) {
```

Comment:
suggestion (non-blocking, but pls do consider 🙏 ): A Go purist would want this constructor to return a concrete type (ie, a `Postgres` struct) rather than an interface, as an application of [the Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle). They would go further and say that the `database` package should not provide interfaces at all, and instead should only provide the concrete struct (ie, the opposite of what has been done here). This way the consumer can pick and define the interface methods that they only care about in an à la carte manner.

You can read more about this in the [official Go wiki](https://github.com/golang/go/wiki/CodeReviewComments#interfaces).

I generally side with the purists here, and I have found that it provides a lot more flexibility and leads to smaller, better-scoped interfaces. It took me a while to digest this though, as seen in our [Slack](https://nilenso.slack.com/archives/C038Z1RHRE0/p1649843312025129) [threads](https://nilenso.slack.com/archives/C036224LCNP/p1662454553220829?thread_ts=1662445589.654249&cid=C036224LCNP).

There are some exceptions within the standard library though, like `http.Handler`. I suppose for something like a web service, I'd want to still expose and return the concrete type.

Reply:
Returning concrete structs makes sense.

References linked here were insightful!